interpreter/go/go.go

```diff
@@ -0,0 +1,170 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package golang // import "go.opentelemetry.io/ebpf-profiler/interpreter/go"
+
+/*
+#cgo CFLAGS: -g -Wall
+#include "../../rust-crates/symblib-capi/c/symblib.h"
+#include <stdlib.h>
+*/
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"runtime"
+	"sync/atomic"
+	"unsafe"
+
+	"go.opentelemetry.io/ebpf-profiler/host"
+	"go.opentelemetry.io/ebpf-profiler/interpreter"
+	"go.opentelemetry.io/ebpf-profiler/libpf"
+	"go.opentelemetry.io/ebpf-profiler/metrics"
+	"go.opentelemetry.io/ebpf-profiler/remotememory"
+	"go.opentelemetry.io/ebpf-profiler/reporter"
+	"go.opentelemetry.io/ebpf-profiler/successfailurecounter"
+)
+
+var (
+	// compiler check to make sure the needed interfaces are satisfied
+	_ interpreter.Data     = &goData{}
+	_ interpreter.Instance = &goInstance{}
+)
+
+type goData struct {
+	exec string
+}
+
+type goInstance struct {
+	interpreter.InstanceStubs
+
+	// Go symbolization metrics
+	successCount atomic.Uint64
+	failCount    atomic.Uint64
+
+	goRuntime *C.SymblibPointResolver
+	pin       runtime.Pinner
+}
+
+func Loader(_ interpreter.EbpfHandler, info *interpreter.LoaderInfo) (
+	interpreter.Data, error) {
+	ef, err := info.GetELF()
+	if err != nil {
+		return nil, err
+	}
+	if !ef.IsGolang() {
+		return nil, nil
+	}
+
+	exec := info.FileName()
+	if len(exec) < 2 {
+		// There are cases where FileName() just returns /.
+		// In these cases we can not access the backing executable
+		// by FileName() and therefore can not continue here.
+		return nil, nil
+	}
+
+	return &goData{
+		exec: exec,
+	}, nil
+}
+
+func (g *goData) Attach(_ interpreter.EbpfHandler, pid libpf.PID,
+	_ libpf.Address, _ remotememory.RemoteMemory) (interpreter.Instance, error) {
+	gi := &goInstance{}
+
+	executablePath := C.CString(fmt.Sprintf("/proc/%d/root%s", pid, g.exec))
+	defer C.free(unsafe.Pointer(executablePath))
```

Comment:
This also should happen in `Loader` *not* in `Attach` so you create a rust symbolizer per Go executable (and not per Go process).

This will prevent coredump testing, as a runtime file is assumed that might not exist. To make coredump testing working, you should not hardcode this path here. I would probably amend `pfelf.Reference` or `pfelf.File` to have a new method that can return the filename, and allow the coredump test suite to hook that call to extract the compressed Go executable as a temporary file.

Reply:
Does something like this work for the test cases https://github.com/open-telemetry/opentelemetry-ebpf-profiler/commit/447b25d65f22a92f3566ffad93558f220f032890 ?

Reply:
Seems no, but https://github.com/fabled/opentelemetry-ebpf-profiler/commit/798b5377217609e529d728a20f9e4fa4890d5e93 should fix it.

I managed to test this locally now. (Seems my kernel produces good cores, but latest gdb does not.)

Reply:
Cherry picked https://github.com/fabled/opentelemetry-ebpf-profiler/commit/798b5377217609e529d728a20f9e4fa4890d5e93.